py_ecc/fields/field_elements.py

```diff
@@ -0,0 +1,355 @@
+        else:
+            return ((self * self) ** int(other // 2)) * self
+
+    # Extended euclidean algorithm used to find the modular inverse
+    def inv(self) -> "FQP":
+        lm, hm = (
+            [1] + [0] * self.degree,
+            [0] * (self.degree + 1),
+        )
+        low, high = (
+            # Ignore mypy yelling about the inner types for  the tuples being incompatible
+            cast(List[IntOrFQ], list(self.coeffs + (0,))),  # type: ignore
+            cast(List[IntOrFQ], list(self.modulus_coeffs + (1,))),  # type: ignore
+        )
+        while deg(low):
+            r = cast(List[IntOrFQ], list(poly_rounded_div(high, low)))
+            r += [0] * (self.degree + 1 - len(r))
+            nm = [x for x in hm]
+            new = [x for x in high]
+            if len(set(
+                [len(lm), len(hm), len(low), len(high), len(nm), len(new), self.degree + 1]
+            )) != 1:
+                raise Exception("Mismatch between the lengths of lm, hm, low, high, nm, new")
+
+            for i in range(self.degree + 1):
+                for j in range(self.degree + 1 - i):
+                    nm[i + j] -= lm[i] * int(r[j])
+                    new[i + j] -= low[i] * int(r[j])
+            lm, low, hm, high = nm, new, lm, low
+        return type(self)(lm[:self.degree], self.curve_name) / low[0]
+
+    def __repr__(self) -> str:
+        return repr(self.coeffs)
+
+    def __eq__(self, other: "FQP") -> bool:     # type: ignore # https://github.com/python/mypy/issues/2783 # noqa: E501
+        if not isinstance(other, type(self)):
+            raise ValueError(
+                "Expected an FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+        for c1, c2 in zip(self.coeffs, other.coeffs):
+            if c1 != c2:
+                return False
+        return True
+
+    def __ne__(self, other: "FQP") -> bool:     # type: ignore # https://github.com/python/mypy/issues/2783 # noqa: E501
+        return not self == other
+
+    def __neg__(self) -> "FQP":
+        return type(self)([-c for c in self.coeffs], self.curve_name)
+
+    @classmethod
+    def one(cls, curve_name: str) -> "FQP":
+        return cls([1] + [0] * (cls.degree - 1), curve_name)
+
+    @classmethod
+    def zero(cls, curve_name: str) -> "FQP":
+        return cls([0] * cls.degree, curve_name)
+
+
+class FQ2(FQP):
+    """
+    The quadratic extension field
+    """
+    degree = 2
+
+    def __init__(self, coeffs: Sequence[IntOrFQ], curve_name: str) -> None:
+        FQ2_MODULUS_COEFFS = field_properties[curve_name]["fq2_modulus_coeffs"]
+        super().__init__(coeffs, curve_name, FQ2_MODULUS_COEFFS)
+
+
+class FQ12(FQP):
+    """
+    The 12th-degree extension field
+    """
+    degree = 12
+
+    def __init__(self, coeffs: Sequence[IntOrFQ], curve_name: str) -> None:
+        FQ12_MODULUS_COEFFS = field_properties[curve_name]["fq12_modulus_coeffs"]
```

Comment:
I don't love that this would just through a `KeyError` if someone used the wrong `curve_name`. I think we'll have that problem in a number of places.

Let's think about alternatives to this setup. Maybe something like:

```py
class Curve(NamedTuple):
  field_modulus: int
  fq12_modulus_coeffs: Tuple[int, ...]
  fq_modulus_coeffs: Tuple[int, int]

bls12_381 = Curve(
  field_modulus = 21888242871839275222246405745257275088696311157297823662689037894645226208583,
  fq2_modulus_coeffs = (1, 0),
  fq12_modulus_coeffs = (2, 0, 0, 0, 0, 0, -2, 0, 0, 0, 0, 0),  # Implied + [1]
)
```

Which makes this method something like:
```py
def __init__(self, coeffs: Sequence[IntOrFQ], curve: Curve) -> None:
  super().__init__(coeffs, curve, curve.fq12_modulus_coeffs)
```
Of course, it comes at the downside of having to import the curve to create an `FQ`. But for the caller, it means fewer magic strings and less likelihood for a typo with a confusing error. For the library, hopefully better readability (at least this method got simpler).

Reply:
@carver is it ok, if I take this up in a future PR?